lib/metasploit/framework/compiler/mingw.rb

```diff
@@ -59,15 +60,54 @@ def build_cmd(src)
           cmd
         end
 
-        def cleanup_files
-          src_file = "#{self.file_name}.c"
-          exe_file = "#{self.file_name}.exe"
 
+        def compile_cpp(src)
+          cmd = build_cpp_cmd(src)
+          print("#{cmd}\n") if show_compile_cmd
+          stdin_err, status = Open3.capture2e(cmd)
+          stdin_err
+        end
+
+        def compile_cpp_files(files)
+          cmd = build_cpp_files_cmd(files)
+          print("#{cmd}\n") if show_compile_cmd
+          stdin_err, status = Open3.capture2e(cmd)
+          stdin_err
+        end
+
+        def build_cpp_cmd(src)
+          src_file = Tempfile.create(['scexec', ".cpp"])
+          File.write(src_file.path, src)
+          build_cpp_files(src_file.path)
+        end
+
+        def build_cpp_files_cmd(file_array)
+          cmd = [ mingw_bin ]
+          cmd << file_array.join(" ")
+          cmd << "-I #{INCLUDE_DIR}"
+          cmd << "-o #{outfile}"
+          cmd << "-shared" if outfile.ends_with?('.dll')
+
+          cmd << opt_lvl if OPTIMIZATION_FLAGS.include?(opt_lvl)
```

Comment:
So my only concern with this is that you end up silently failing vs alerting the user they passed in an incorrect option. This may lead to users compiling something and thinking it has been appropriately optimized when in fact it has not been.

Open to people's opinions on this though.

Reply:
This should only add the option  if it exists in the list of accepted optimization flags. Did I understand your comment correctly?

Reply:
I guess my original thought here was to provide the user more feedback if they passed in an incorrect optimization flag but now that I review this comment, it really doesn't make much sense. Most compilers will also just ignore incorrect flags if they are passed in.

Really my original thought was to prevent a case where a user passes in some high optimization value like say 6 when we only go up to optimization level 5, and then he is mislead into thinking his binary was highly optimized when in reality it was not optimized at all.